### PR TITLE
[Enhancement #35636] Allow passing variables to `wc_get_template_part`

### LINKED ISSUE #35636: [Enhancement]: wc_get_template_part passing variable
### Describe the solution you'd like

`wc_get_template_part` should accept the third param as passing variable. Simply make its interface looks like [the core function](https://developer.wordpress.org/reference/functions/get_template_part).

### Describe alternatives you've considered

I have to use `global $product` which I usually try to avoid in my work.

### Additional context

I think I can do this. Please assign me this task.

### PR DESCRIPTION
**Pull request by ngdangtu-vn**
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

#### Changing explaination

- Allow passing variables to `wc_get_template_part` through the third param `$args`.
- Refactor code, including rename variable and update comments. It serves readable purpose.

### Changes proposed in this Pull Request:

Closes #35636 . The reason is explained in issue 35636.

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

1.
2.
3.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by ngdangtu-vn** [comment]
Hello, this is my first PR to the project and PHP is not my first-tier language so I may do things wrong. If so, please kindly tell me what I need to fix, I do my best to fulfill PR requirement.

For now, I will submit draft PR first. Later on I will update changelogger use. I don't know how to do the test, but I will do research on it as well. However, it may take sometime because I'm not able to setup Woocommerce environment yet.


### Update

- About the testing, even though it is a small update, but I also did refactor so I think a test for double checking should be consider. I just don't know how yet.
- I will rebase all the commits into one before I do final PR. So if you need to pull my repo, please be careful.

**Comment by jorgeatorres** [comment]
Hi @ngdangtu-vn!

Thanks for contributing to WooCommerce! 🥇

While this seems to improve consistency with WP, keep in mind that `$args` is rarely used in WP itself for template parts. Also, this new `$args` parameter will be inconsistent with our own `wc_get_template()`, which might be confusing: notice that `wc_get_template()` uses `extract()` on the passed `$args`, so the template file gets all those array keys/values as variables on their own, while, in this case, `wc_get_template_part()` would just pass one variable (`$args`) to the template.

Before we introduce a new parameter (with the associated compatibility implications), I'd like to better understand your use case here: are you frequently passing vars to template parts? in which contexts?
Our default templates don't even make much use of template parts (except when already global objects are involved) which is why I'm asking. Maybe your use-case might be better served by other function.

In any case, I'm not necessarily opposing the change, but I do think we should discuss before moving forward.

Please let me know what you think. And once again, thanks for contributing!

**Comment by ngdangtu-vn** [comment]
> While this seems to improve consistency with WP, keep in mind that `$args` is rarely used in WP itself for template parts... I'd like to better understand your use case here... are you frequently passing vars to template parts? in which contexts?

I wouldn't said I always need to use it, but I think it would help me to make the template parts more flexible. My latest use case is passing a small boolean toggle to the `rating.php`, sometime I need to make it display more info and the other time less info.


> Also, this new `$args` parameter will be inconsistent with our own `wc_get_template()`, which might be confusing: notice that `wc_get_template()` uses `extract()` on the passed `$args`, so the template file gets all those array keys/values as variables on their own, while, in this case, `wc_get_template_part()` would just pass one variable (`$args`) to the template.

I had checked the `wc_get_template()` case as well. And I think if we talk about consistency in code logic, we are both right and wrong. Wrong is because `wc_get_template()` uses `include` to load a template inside its own function while `wc_get_template_part()` use the wp core function ([load_template](https://developer.wordpress.org/reference/functions/load_template/#source)) to load the template. And the right thing is the `load_template()` also uses `extract()`.



So if you are worry about inconsistency of the functions' logic, I can reassure you it'll be fine. The only thing I'm afraid is me or someone else may misuse the feature and make the website slow down (I got one due to functional style and passing a big var through function).

---

Btw, how do you think about my little refactor for variables in the function? I was confused by
reusing name for variable so I thought I could make it more readable. But I'm worry if it could break the code since I couldn't test it right now.

So in the end, I hope I have cleared out the problems we are having? (and accidentally introduce new ones). Thank you very much for your question, it sure helps me be more careful and double think about the code I did.

Please tell me what you think about this feature, my use case and the refactoring.

**Comment by jorgeatorres** [comment]
Happy new year @ngdangtu-vn!

> I had checked the wc_get_template() case as well. And I think if we talk about consistency in code logic, we are both right and wrong. Wrong is because wc_get_template() uses include to load a template inside its own function while wc_get_template_part() use the wp core function ([load_template](https://developer.wordpress.org/reference/functions/load_template/#source)) to load the template. And the right thing is the load_template() also uses extract().

I agree that WC's usage is inconsistent with WordPress'. I'm not suggesting either approach is "the right thing" but I do worry that this change to `wc_get_template_part()` would mean that we're introducing an inconsistency with our own `wc_get_template()` function.

Let me explain:

- For the sake of argument, say you want to pass `array( 'myvar' => 1 )` as `$args` to either function.
- As you said so yourself, `wc_get_template()` uses `extract()` on the `$args` so you can actually refer to `$myarg` directly inside the template file.
- `wc_get_template_part()`, on the other hand, uses `load_template()`... which seems like an improvement, but actually, `load_template()` doesn't really call `extract()` on `$args`.
The variables it makes available via `extract()` [are the query vars in the main query](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/template.php#L764). This means, any template file loaded via `wc_get_template_part()` would have access to `$args` itself but not to `$myvar` directly.

I actually prefer that we don't call `extract()` at all, so I like your approach, but I do worry about the inconsistency, even if `$args` are seldom used. I mean, from a dev perspective, you wouldn't expect that `wc_get_template()` and `wc_get_template_part()` would work so differently, but we can't change `wc_get_template()` at this point. We can, instead, make `wc_get_template_part()` compatible but that might be too much effort.

I want to discuss this with the team. In the mean time, and to make reviewing easier would you mind keeping this PR focused on the introduction of `$args` and save the refactoring/renaming (which I find useful) for a future PR? Also, we need to add `$args` to the docblock for `wc_get_template_part()`.

Please let me know what you think.

**Comment by ngdangtu-vn** [comment]
Happy New Year to you too, @jorgeatorres.

You are right. It's my bad for mistaken the `load_template()`. Now I totally understand your worry. The bad news seems like we cannot change the `wc_get_template()`... But if we need to follow the `extract()` way, it means that we will have to load template without the core function. Anyway, I think it is better to wait for you and your team make the decision. I don't mind to wait at all.

About the refactor/rename variables, I agree that we should have this as separate PR. But could I have the refactor PR to be merged before this PR? If possible of course. I think it will be easily to work with once we have a better name.

In short, I agree with your opinion.

Btw, here is my new branch for renaming: https://github.com/ngdangtu-vn/woocommerce/tree/refactor-wc_get_template_part

**Comment by barryhughes** [comment]
> The bad news seems like we cannot change the wc_get_template()... But if we need to follow the extract() way, it means that we will have to load template without the core function.

Exactly, and unfortunately I do not believe we can stop using `load_template()` without introducing a different set of backwards compatibility concerns (there may be existing code depending on hooks like `wp_before_load_template` ... and I'd be reluctant to replicate them given the difference in how we'd be handling `$args`).

With that in mind, I'll go ahead and close this PR. Please don't hesitate to reach out if you want to discuss further ... but, unfortunately, this approach doesn't seem viable.
